File: Programmers/[17685] 자동완성/JFe/17685.py
from collections import defaultdict
def solution(words):
    answer = 0
    length = defaultdict(int)
    words.sort()
    for i in range(len(words)-1):
        count = 0
        # 뒷 단어와 공통된 문자 수 체크
        for j in range(len(words[i])):
            if words[i][j] != words[i+1][j]: break
            count += 1
        # 원래 단어 수와 같으면 count 그대로
        if count == len(words[i]):
            length[i] = max(length[i], count)
        # 원래 단어 수와 다르면 count + 1
        else:
            length[i] = max(length[i], count+1)
        # 원래 단어 수와 같으면 count 그대로
        if count == len(words[i+1]):
            length[i+1] = count
        # 원래 단어 수와 다르면 count + 1
        else:
            length[i+1] = count+1
    return sum(length.values())

File: Programmers/[17685] 자동완성/woong-jae/17685.js
class Node {
    constructor() {
        this.count = 0;
        this.childs = new Map();
    }
}

function solution(words) {
    let answer = 0;
    const head = new Node();
    
    // 트라이 구성
    words.forEach(word => {
        let cur = head, depth = 1;
        while(depth <= word.length) {
            const next = word.slice(0, depth);
            if(!cur.childs.has(next)) {
                cur.childs.set(next, new Node());
            }
            cur = cur.childs.get(next);
            cur.count++;
            depth++;
        }
    });
    
    // 입력해야 할 문자수 구하기
    words.forEach(word => {
        let cur = head, depth = 1;
        while(depth < word.length) {
            const next = word.slice(0, depth);
            cur = cur.childs.get(next);
            if(cur.count === 1) break;
            depth++;
        }
        answer += depth;
    });
    
    return answer;
}
